### PR TITLE
Configurable feature file reader

### PR DESCRIPTION
**Pull request by marcin-jelenski**
Flutter has an [integration_test](https://pub.dev/packages/integration_test) package for E2E testing, which does not have an access to host device files and possibly can replace flutter driver in the nearest future.

Let's remove glob library dependency, expose interfaces for feature files indexing and reading so a platform/device/framework-specific implementation could be used.

All tests have passed. The feature is backwards-compatible with `Glob` patterns on the previous version.

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -1,10 +1,9 @@
 name: gherkin
-version: 1.1.9
+version: 2.0.0
```
Comment (marcin-jelenski) [inline pubspec.yaml:2]: As this is breaking feature, we should consider increasing major version.

**Comment by marcin-jelenski** [comment]
CC @jonsamwell

**Comment by marcin-jelenski** [comment]
And here's the working config for flutter with integration_tests package (candidate for the separate flutter package):

```dart
void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets("gherkin tests", (WidgetTester tester) async {
    await tester.runAsync(() async {
      await tester.pumpWidget(MyApp());
      await tester.pumpAndSettle();
    });

    final config = TestConfiguration()
      ..features = [
        RegExp(r"features/.*\.feature"),
      ]
      ..reporters = [
        StdoutReporter(MessageLevel.error),
        ProgressReporter(),
        TestRunSummaryReporter()
      ]
      ..featureFileIndexer = FlutterFeatureFilesIndexer()
      ..featureFileReader = FlutterFeatureFileReader()
      ..createWorld = (TestConfiguration config) {
        return Future.value(GherkinWorld(tester));
      }
      ..stepDefinitions = stepDefinitions
      ..exitAfterTestRun = true;

    await GherkinRunner().execute(config);
  });
}

class GherkinWorld extends World {
  final WidgetTester tester;

  GherkinWorld(this.tester);
}

class FlutterFeatureFileReader implements FeatureFileReader {
  @override
  Future<String> readAsString(String path) async {
    return await rootBundle.loadString(path);
  }
}

class FlutterFeatureFilesIndexer implements FeatureFileIndexer {
  @override
  Future<List<String>> listFiles(Pattern pattern) async {
    final manifest = await rootBundle.loadString('AssetManifest.json');
    final Map<String, dynamic> manifestMap = jsonDecode(manifest);

    return manifestMap.keys
        .where((element) => pattern.allMatches(element).isNotEmpty)
        .toList();
  }
}
```

**Comment by jonsamwell** [comment]
@marcin-jelenski thank you for this it is excellent work, this will be a key step in the null-safety conversion as well.

Regarding the `integration _test` package there is a working branch in flutter_gherkin that has this working and I will very soon be publishing this branch to master and publishing it.  It relies on code generation at development time to turn the feature files into code rather than having to include the files as app assets.  It also relies on `TestWidget` rather than Flutter driver as this is the way the Flutter team now suggest integration testing.  See https://github.com/jonsamwell/flutter_gherkin/blob/integration_test__package_support/CHANGELOG.md

I'll try and get this merged in tomorrow

**Comment by marcin-jelenski** [comment]
@jonsamwell thank you for the code generation tip. I've evaluated it on a few target platforms (Android, iOS, Web) and seems to work seamlessly. Said that I will remove Readme examples with flutter assets loading.

File: lib/src/test_runner.dart [hunk 2]
```diff
@@ -96,9 +97,8 @@ class GherkinRunner {
     }
 
     await _reporter.dispose();
-    exitCode = allFeaturesPassed ? 0 : 1;
 
-    if (config.exitAfterTestRun) exit(allFeaturesPassed ? 0 : 1);
+    if (config.exitAfterTestRun && allFeaturesPassed) throw 'Tests failed';
```
Comment (jonsamwell) [inline lib/src/test_runner.dart:101]: @marcin-jelenski what was the rationale behind this change?

Comment (marcin-jelenski) [reply lib/src/test_runner.dart:101]: @jonsamwell as mentioned above, `dart:io` import in test_runner breaks compatibility with dartjs and breaks tests in dart web-related apps (not flutter-one). There's no `exitCode` there. We're using gherkin for a variety of `angular_dart` tests and that line was a blocker. If you want, I can revert this and prepare a separate PR for this.

**Pull request merged**